R/epichains.R

```diff
@@ -0,0 +1,283 @@
+
+    num_generations <- length(unique(object$generation))
+
+    max_generation <- max(object$generation)
+
+    # out of summary
+    res <- list(
+      unique_chains = chains_ran,
+      max_time = max_time,
+      unique_ancestors = n_unique_ancestors,
+      unique_generations = n_unique_ancestors,
+      num_generations = num_generations,
+      max_generation = max_generation
+    )
+  } else if (attributes(object)$chain_type == "chains_vec") {
+    chains_ran <- length(object)
+    max_chain_stat <- max(!is.infinite(object))
+    min_chain_stat <- min(!is.infinite(object))
+
+    res <- list(
+      unique_chains = chains_ran,
+      max_chain_stat = max_chain_stat,
+      min_chain_stat = min_chain_stat
+    )
+  }
+
+  return(res)
+}
+
+#' Checks whether the object is an `epichains`
+#'
+#' @param x An R object
+#'
+#' @return logical, `TRUE` if the object is an `epichains` and `FALSE`
+#' otherwise
+#' @keywords internal
+is_epichains <- function(x) {
+  inherits(x, "epichains")
+}
+
+#' Check if an object is of class "epichains_aggregate_df"
+#'
+#' @param x An [`epichains`] object
+#'
+#' @keywords internal
+is_epichains_aggregate_df <- function(x) {
+  if (!inherits(x, "epichains_aggregate_df")) {
+    stop("Object must have class 'epichains_aggregate_df'")
+  }
+}
+
+#' `epichains` class validator
+#'
+#' @param x An `epichains` object
+#'
+#' @return Checks if an object is of class `epichains` and if so
+#' checks that it's in the right format as a "data.frame" or vector.
+#' @keywords internal
+#' @author James M. Azam
+validate_epichains <- function(x) {
+  if (!is_epichains(x)) {
+    stop("Object must have an epichains class")
+  }
+
+  # check for class invariants
+
+  if (attributes(x)$chain_type == "chains_tree") {
+    stopifnot(
+      "object does not contain the correct columns" =
+        c("sim_id", "ancestor", "generation", "time") %in%
+          colnames(x),
+      "column `sim_id` must be a numeric" =
+        is.numeric(x$sim_id),
+      "column `ancestor` must be a numeric" =
+        is.numeric(x$ancestor),
+      "column `generation` must be a numeric" =
+        is.numeric(x$generation),
+      "column `time` must be a numeric" =
+        is.numeric(x$time)
+    )
+  } else {
+    stopifnot(
+      "object must be a numeric vector" =
+        is.numeric(x)
+    )
+  }
+
+  invisible(x)
+}
+
+#' `head` method for [`epichains`] class
+#'
+#' @param x An [`epichains`] object
+#' @param ... further arguments passed to or from other methods
+#' @importFrom utils head
+#' @return object of class `data.frame`
+#' @author James M. Azam
+#' @export
+head.epichains <- function(x, ...) {
+  utils::head(as.data.frame(x), ...)
+}
+
+#' `tail` method for [`epichains`] class
+#' @param x An [`epichains`] object
+#' @param ... further arguments passed to or from other methods
+#' @importFrom utils tail
+#' @author James M. Azam
+#' @export
+tail.epichains <- function(x, ...) {
+  utils::tail(as.data.frame(x), ...)
+}
```

Comment:
Is an epichains object also a data.frame object (or could it also be)? If so, I *think* these should be unnecessary?

Reply:
Yes, it is also a `data.frame`. However, I have been considering implementing a `head` method that only prints from the first known ancestor.